File: csrc/tests/test_identity_parity.cpp
// The identity port, held to the REFERENCE's answer rather than to my own expectations.
//
// `benchmarks/parity/scenarios/identity/basic.txt` is read by both planes;
// `benchmarks/parity/golden/identity/basic.txt` is what `shipvision.mtmc.identity` answered to
// it, emitted by `scripts/emit_parity_golden.py --kind identity`. This binary replays the
// scenarios through `GlobalIdAssigner` and compares line for line -- so a change to either
// implementation that moves one id fails here, which is the only thing that keeps a
// five-hundred-line port a port. `test_mtmc_identity.cpp` tests the same class against its
// own reasons; this one tests it against the other plane.
#include <algorithm>
#include <cstdio>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

#include "shipinfer/core/types.h"
#include "shipinfer/pipeline/mtmc/identity.h"
#include "tests/parity_files.h"

namespace {

    using namespace shipinfer;
    using shipinfer::mtmc::GlobalIdAssigner;
    using shipinfer::mtmc::IdentityObservation;
    using shipinfer::mtmc::TrackKey;
    using shipinfer::parity::resolve;

    int failures = 0;
    int checks = 0;

    void check(bool condition, const std::string& what) {
        ++checks;
        if (!condition) {
            ++failures;
            std::printf("FAIL: %s\n", what.c_str());
        }
    }

    std::vector<std::string> lines_of(const std::string& path, bool keep_comments) {
        std::ifstream file(path);
        if (!file) throw ConfigError("cannot read " + path);
        std::vector<std::string> out;
        std::string line;
        while (std::getline(file, line)) {
            while (!line.empty() && (line.back() == ' ' || line.back() == '\r'))
                line.pop_back();
            if (line.empty()) continue;
            if (!keep_comments && line[0] == '#') continue;
            out.push_back(line);
        }
        return out;
    }

    struct Instant {
        std::vector<int> labels;
        std::vector<IdentityObservation> observations;
    };

    struct Scenario {
        std::string name;
        GlobalIdAssigner::Options options;
        std::vector<Instant> instants;
    };

    // `<label>:<camera>#<track>:<x>,<y>` -- the scenario file's own header states the format.
    void parse_field(const std::string& field, Instant& instant) {
        const size_t first = field.find(':');
        const size_t second = field.find(':', first + 1);
        const size_t hash = field.find('#');
        const size_t comma = field.find(',', second);
        if (first == std::string::npos || second == std::string::npos ||
            hash == std::string::npos || comma == std::string::npos) {
            throw ConfigError("malformed observation '" + field + "'");
        }
        instant.labels.push_back(std::stoi(field.substr(0, first)));
        const std::string camera = field.substr(first + 1, hash - first - 1);
        const int64_t track = std::stoll(field.substr(hash + 1, second - hash - 1));
        const float x = std::stof(field.substr(second + 1, comma - second - 1));
        const float y = std::stof(field.substr(comma + 1));
        instant.observations.push_back(IdentityObservation{TrackKey{camera, track}, {x, y}});
    }

    std::vector<Scenario> read_scenarios(const std::string& path) {
        std::vector<Scenario> scenarios;
        for (const std::string& line : lines_of(path, false)) {
            std::istringstream stream(line);
            std::string head;
            stream >> head;
            if (head == "scenario") {
                Scenario scenario;
                stream >> scenario.name >> scenario.options.max_age >>
                    scenario.options.capacity >> scenario.options.max_tracks;
                // ON for the gate: the reference ran with it on, so a state the invariant
                // rejects would be a difference the ids might not show.
                scenario.options.validate_every_step = true;
                scenarios.push_back(std::move(scenario));
                continue;
            }
            if (head != "instant") throw ConfigError("unknown line '" + line + "'");
            if (scenarios.empty()) throw ConfigError("an `instant` before any `scenario`");
            Instant instant;
            std::string field;
            while (stream >> field) parse_field(field, instant);
            scenarios.back().instants.push_back(std::move(instant));
        }
        return scenarios;
    }

    void the_port_answers_what_the_reference_answered() {
        const std::vector<Scenario> scenarios =
            read_scenarios(resolve("scenarios/identity/basic.txt"));
        const std::vector<std::string> golden =
            lines_of(resolve("golden/identity/basic.txt"), false);

        std::vector<std::string> written;
        for (const Scenario& scenario : scenarios) {
            GlobalIdAssigner assigner(scenario.options);
            written.push_back("scenario " + scenario.name);
            for (const Instant& instant : scenario.instants) {
                const auto result = assigner.assign(instant.observations, instant.labels);
                std::string answers;
                for (const auto& [key, global_id] : result) {
                    if (!answers.empty()) answers += " ";
                    answers += key.str() + "=" + std::to_string(global_id);
                }
                written.push_back(answers.empty() ? "result" : "result " + answers);
            }
            written.push_back("issued " + std::to_string(assigner.issued()));
        }

        check(!scenarios.empty(), "the scenario file holds scenarios");
        check(written.size() == golden.size(),
              "the port answered " + std::to_string(written.size()) + " lines and the " +
                  "reference " + std::to_string(golden.size()));
        const size_t common = std::min(written.size(), golden.size());
        size_t agreed = 0;
        for (size_t i = 0; i < common; ++i) {
            if (written[i] == golden[i]) {
                ++agreed;
                continue;
            }
            check(false, "line " + std::to_string(i + 1) + " differs");
            std::printf("  reference: %s\n  port:      %s\n", golden[i].c_str(),
                        written[i].c_str());
        }
        check(agreed == common && common == golden.size(),
              "every line of the reference's answer is the port's answer too (" +
                  std::to_string(agreed) + "/" + std::to_string(golden.size()) + ")");
    }

    void the_golden_names_its_own_emitter() {
        // A golden nobody can regenerate is a golden that gets hand-edited the first time it
        // fails, which is how a gate becomes a rubber stamp.
        const std::vector<std::string> all =
            lines_of(resolve("golden/identity/basic.txt"), true);
        bool named = false;
        for (const std::string& line : all) {
            if (line.find("emit_parity_golden.py --kind identity") != std::string::npos) {
                named = true;
            }
        }

        check(named, "the golden carries the command that emits it");
    }

}  // namespace

int main() {
    try {
        the_port_answers_what_the_reference_answered();
        the_golden_names_its_own_emitter();
    } catch (const std::exception& error) {
        std::printf("FAIL: %s\n", error.what());
        ++failures;
        ++checks;
    }
    std::printf("%d checks, %d failure(s)\n", checks, failures);
    return failures == 0 ? 0 : 1;
}

File: tests/pipeline/test_identity_parity.py
"""The Python half of the identity golden: is the committed file still what the reference says?

The C++ binary (`csrc/tests/test_identity_parity.cpp`) is the real gate -- it is the port that
can drift. This one answers the other question, which that binary cannot: whether the GOLDEN
still matches the reference it was emitted from, so a submodule bump that changes an id is
caught here rather than showing up as "the port is wrong".

Skipped without the submodule, and CI's `Tests (with the kernels' Python half)` job is where it
runs.
"""

from __future__ import annotations

from importlib.util import find_spec
from pathlib import Path

import pytest

from benchmarks.parity.drive_identity import GOLDEN, SCENARIOS, load, render_identity

pytestmark = pytest.mark.skipif(
    find_spec("shipvision") is None,
    reason="the shipvision submodule is not importable; run "
    "`git submodule update --init 3rdparty/shipvision` and put it on PYTHONPATH",
)

#: Every scenario file with a golden. One today; the loop is what keeps the next one covered.
NAMES = tuple(sorted(path.stem for path in SCENARIOS.glob("*.txt")))


class TestTheGoldenIsStillTheReferencesAnswer:
    def test_there_is_a_scenario_at_all(self) -> None:
        """A glob that finds nothing would make every test below vacuously green."""
        assert NAMES, f"no scenario files under {SCENARIOS}"

    @pytest.mark.parametrize("name", NAMES)
    def test_the_golden_is_reproduced_exactly(self, name: str) -> None:
        expected = (GOLDEN / f"{name}.txt").read_text(encoding="utf-8")

        assert render_identity(load(name)) == expected, (
            f"the ids the reference assigns for {name} are not the committed golden. If a "
            f"submodule bump IS the decision, re-emit with `python "
            f"scripts/emit_parity_golden.py --kind identity --scenario {name} --emit-golden "
            f"--force`, say so in the PR, and check the C++ gate in the same commit"
        )

    @pytest.mark.parametrize("name", NAMES)
    def test_the_golden_names_the_command_that_emits_it(self, name: str) -> None:
        """The C++ gate asserts this too. Both, because a golden nobody can regenerate is one
        that gets hand-edited the first time it fails."""
        text = (GOLDEN / f"{name}.txt").read_text(encoding="utf-8")

        assert "emit_parity_golden.py --kind identity" in text


class TestAContestedCameraSlotHasOneHolder:
    """What `MTMC-ONE-CAMERA-TWICE-IN-A-CONTESTED-CLUSTER` was, now asserted the right way up.

    It used to be a tripwire: the reference adopted BOTH challengers and this file pinned that,
    so the day upstream fixed it something would notice. shipvision#17 is that day. The tests
    below are the same two instants with the assertions inverted, and they stay because the
    parity golden reaches the answer but not the reference's own `validate()`.
    """

    def observation(self, camera: str, track: int, x: float, y: float):
        import numpy as np
        from shipvision.mtmc.frames import TrackKey, TrackObservation
        from shipvision.types import FrameTag, Track

        return TrackObservation(
            key=TrackKey(camera_id=camera, track_id=track),
            track=Track(
                track_id=track,
                box=np.array([0.0, 0.0, 100.0, 300.0]),
                tag=FrameTag(camera_id=camera, frame_id=0),
                embedding=np.array([x, y], dtype=np.float32),
            ),
            frame_height=1080,
            frame_width=1920,
        )

    def _contested(self, validate: bool, second: float, third: float):
        from shipvision.mtmc.identity import GlobalIdAssigner

        look = self.observation
        assigner = GlobalIdAssigner(validate_every_step=validate)
        assigner.assign(
            [look("cam0", 1, 0, 1), look("cam1", 1, 1, 0), look("cam2", 1, 1, 0)], [0, 0, 0]
        )
        assigner.assign(
            [
                look("cam0", 1, 0, 1),
                look("cam1", 1, 1, 0),
                look("cam2", 1, 1, 0),
                look("cam0", 2, 1, second),
                look("cam0", 3, 1, third),
            ],
            [1, 0, 0, 0, 0],
        )
        return assigner

    @pytest.mark.parametrize(
        "second,third,holder", [(0.3, 0.8, 2), (0.8, 0.3, 3)], ids=["second", "third"]
    )
    def test_only_the_better_challenger_holds_the_slot(
        self, second: float, third: float, holder: int
    ) -> None:
        """Both directions. A fix that skipped the displaced track without still holding a
        contest would pass the first case and fail the second."""
        from shipvision.mtmc.frames import TrackKey

        # `validate_every_step=False` is production's default, and how this stayed silent.
        assigner = self._contested(False, second, third)

        target = assigner.owner_of(TrackKey(camera_id="cam1", track_id=1))
        from_cam0 = [key for key in assigner.members(target) if key.camera_id == "cam0"]

        assert from_cam0 == [TrackKey(camera_id="cam0", track_id=holder)], (
            f"a contested camera slot has one holder, and it is the better challenger. "
            f"{len(from_cam0)} here means the upstream fix has been reverted or lost in a "
            f"submodule bump; the C++ port in `csrc/.../mtmc/identity.cpp` carries the same "
            f"rule and `golden/identity/basic.txt` holds the answer"
        )

    def test_the_reference_validates_its_own_state_now(self) -> None:
        """The half no golden can carry: asked to check itself, the reference used to call the
        state it had just built a fault. `validate_every_step=True` is what the parity driver
        runs with, so this is also why the golden can hold these scenarios at all."""
        self._contested(True, 0.3, 0.8)  # raises TrackingError if the invariant is broken


class TestTheCppGateReadsTheSameFiles:
    def test_the_binary_names_this_scenario_and_this_golden(self) -> None:
        """A path typo would leave the C++ gate reading a file nobody emits, and it would pass.
        Source-level because a Python test cannot run the binary."""
        source = (
            Path(__file__).resolve().parents[2] / "csrc" / "tests" / "test_identity_parity.cpp"
        ).read_text(encoding="utf-8")

        for name in NAMES:
            assert f"scenarios/identity/{name}.txt" in source
            assert f"golden/identity/{name}.txt" in source
